File: DesignPatternsTutorial/singleton.php
<?php

class DataBase
{

    private static $instance = null; //data type: DataBase
    private $connection;
    private $user;
    private $password;
    private $dbName;

    private function __construct($c, $u, $p, $db)
    {
        $this->connection = $c;
        $this->user = $u;
        $this->password = $p;
        $this->dbName = $db;
    }

    public static function getInstance(): DataBase
    {
        $currentInstance = self::$instance;
        if (!isset($currentInstance)) {
            self::$instance = $currentInstance = new DataBase('localhost', 'root', '', 'test');
        }

        return $currentInstance;
    }

}


function main()
{
    $db1 = DataBase::getInstance();
    $db2 = DataBase::getInstance();
    if ($db1 === $db2) {
        echo "Singleton works, both variables contain the same instance.";
    } else {
        echo "Singleton failed, variables contain different instances.";
    }
}

main();

File: DesignPatternsTutorial/singleton.cpp
#include <iostream>
using namespace std;

class Database {
private:
    static Database* instance;

    Database() {
        cout << "Database initialized!" << endl;
    }

public:
    static Database& getInstance() {
        if (instance == nullptr)
            instance = new Database();
        return *instance;
    }
};

Database* Database::instance = nullptr;

int main() {
    Database& db1 = Database::getInstance();
    Database& db2 = Database::getInstance();

    cout << "db1 address: " << &db1 << endl;
    cout << "db2 address: " << &db2 << endl;

    if (&db1 == &db2) {
        cout << "Both instances are the same!" << endl;
    } else {
        cout << "Instances are different!" << endl;
    }

    return 0;
}
